Ray.jl/src/function_manager.jl

```
@@ -0,0 +1,78 @@
+    gcs_client::JuliaGcsClient
+    functions::Dict{String,Any}
+end
+
+# TODO: remove indirection after fields are settled
+FunctionManager = FunctionManager1
+
+# XXX: we should probably be packaging task-related metadata like the job_id in
+# a task/remotefunction-like struct instead of passing it around like this.
+function_key(fd::JuliaFunctionDescriptor, job_id) = string("RemoteFunction:", job_id, ":", fd.function_hash)
+
+function export_function!(fm::FunctionManager, f, job_id)
+    fd = function_descriptor(f)
+    key = function_key(fd, job_id)
+    if Exists(fm.gcs_client, FUNCTION_MANAGER_NAMESPACE, key, -1)
+        @debug "function already present in GCS store:" key f
+    else
+        @debug "exporting function to GCS store:" key f
+        val = base64encode(serialize, f)
+        Put(fm.gcs_client, FUNCTION_MANAGER_NAMESPACE, key, val, true, -1)
+    end
+end
+
+function wait_for_function(fm::FunctionManager, fd::JuliaFunctionDescriptor, job_id;
+                           pollint_s=0.01, timeout_s=10)
+    key = function_key(fd, job_id)
+    status = timedwait(timeout_s; pollint=pollint_s) do
+        # timeout the Exists query to the same timeout we use here so we don't
+        # deadlock.
+        Exists(fm.gcs_client, FUNCTION_MANAGER_NAMESPACE, key, timeout_s)
+    end
+    return status
+end
+
+# XXX: this will error if the function is not found in the store.
+function import_function!(fm::FunctionManager, fd::JuliaFunctionDescriptor, job_id)
+    return get!(fm.functions, fd.function_hash) do
+        key = function_key(fd, job_id)
+        @debug "retrieving $(fd) from function store with key $(key)"
+        val = Get(fm.gcs_client, FUNCTION_MANAGER_NAMESPACE, key, -1)
+        try
+            io = IOBuffer()
+            iob64 = Base64DecodePipe(io)
+            write(io, val)
+            seekstart(io)
+            f = deserialize(iob64)
+        catch e
+            error("Failed to deserialize function from store: $(fd)")
```

maybe rethrow the error?